Feature/jbm/module10

Jiymeade opened the pull request:
I've made changes based on the last comments. Although it's better, there are still issues in unity.

Jiymeade committed:
Attempt of module 10

Minimal Requirements

Adjust ServerPlayerMovement's movement and add fire controls talking to bullet spawner
Bullet Spawner system attached to serverPlayerPrefab
Projectile and handling collision
Health System with simple image slider to update and sync
Pickup example using ammo and a simple spawner in the scene
Very basic Game Logic

Jiymeade committed:
Changes

-Module 10 changed made based off of feedback.. There is still issues suck as the losing text not showing up and a repeated use of explosion also the character floating.

MRN-DX commented:
I'm looking at it now, i'll post comments as I go by.

So I see you've been modifying these game objects for spawning and testing etc.
![image](https://github.com/user-attachments/assets/9bd80298-b28d-4d2a-a701-41ea8a18c650)


However, if you do you have to remember to Override the base prefab. To do this from scene, you want to click on the object thats tied to a prefab, in this case Explosion. Click override and here you can either pick and choose which change to revert to the base prefab (in case you make a mistake) or apply to the base prefab so it affects any duplicates, or in our case updates the object we want to spawn. If its the most up-to-date prefab, then quickest way is to hit apply all. 

![image](https://github.com/user-attachments/assets/aeab2c0e-7374-4b89-94df-c7dd501bb61c)

Usually when I edit base items, I like to open the prefab on its own so I don't have to worry about overrides, but that's just my personal taste.


![image](https://github.com/user-attachments/assets/d97bf166-2189-467e-973e-2e95c7448659)

So when I dove into Explosions prefab you can see that its missing the animator and new additions. that's because the override has been applied. The networkManger talks to the ones in the assets so you would expect to see bugs or missing items than the ones you see in game.
![image](https://github.com/user-attachments/assets/4f444d27-6ad5-498a-92d9-e8dcf620c05a)

you would definitely want to update all your prefabs

MRN-DX commented:
You're spawners need to reference the asset prefab on file, not in scene for it to work properly, Click and drag them onto the script. Anything that we are creating like spawning dynamically we will always want to point to the one in our resources.
![image](https://github.com/user-attachments/assets/ce4c8aec-97f2-4505-93cc-f7d29becfffd)

MRN-DX commented:
Missing reference on Base TnT script for explosion.
![image](https://github.com/user-attachments/assets/f9d247f2-8b62-4984-b422-2f0a3791acf0)

The explosion animation motion is missing an Animation event in its timeline
![image](https://github.com/user-attachments/assets/52399d80-4ab5-4dd1-9284-3fa0a448b8b7)

then make sure to set the method it needs to talk to from the inspector, also note you must be inside the base explosion prefab for this change to work right.
![image](https://github.com/user-attachments/assets/f6e459bd-e80e-4309-97d0-11b39067c994)

The last thing explosion prefab needs is the missing rigibody with wthe X,Y,Z, positions frozen, but the scale xyz unfrozen. gravity off.

MRN-DX commented:
Here's a quick way to reposition anything in Unity, I like to use the GameObject Tab up top > Move To View, when I want to move something. So for example using your spawners as example, lets say I want them to be near the playing field.... 

I first focus the camera to the ground by double clicking it in the scene's hagiarchy, the camera should move.
![image](https://github.com/user-attachments/assets/a2a771e0-a1d6-4fc7-a32f-e8f3ef7979d3)

Then I single click the target I want, in this case the root of your spawners, and click move to view
![image](https://github.com/user-attachments/assets/ba655ee6-4130-4bd2-aaf7-c5a80de59ff5)

![image](https://github.com/user-attachments/assets/42c11bc0-1c06-4d49-9ba0-aaf1354064bd)

then all you need to do is some easy tweaking
![image](https://github.com/user-attachments/assets/2b60dc42-e180-4275-ac92-e73445e55d0b)

MRN-DX commented:
ServerPlayerPrefab RigiBody has frozen X and Z positions, need to unfreeze them by unchecking the boxes
![image](https://github.com/user-attachments/assets/5222abc5-ac2e-41a9-9558-f372f787d221)

MRN-DX commented:
Some other tweaks that I would do to make the physics a little better would be to increase the mass of the ground dramatically, increase the angular drag if you want it to be less spinny, but those are some minor tweaks considering.

Jiymeade committed:
Changes pt 2.

-The ammo pickup isn't active
-Explosion issue in only one spot
-My WinText isn't showing up when the game starts but the message plays if you lose and look at the scene, though it doesn't play the right text when the player dies

Jiymeade commented:
-I put the ammo pick up into the Tnt spawner for 2 of the Tntspawners and It spawns. So, that issue is fixed and the player can pick them up but now it seems like its a problem with the collision of objects with the ground plane because it's like a item floating in water physic about it.

Jiymeade committed:
Small Changes

- Fixed the issue of the explosion not happening but It happens not where the tnt is spawning
-The Bullets and TNT are sinking into the ground Plane

Jiymeade commented:
Please look at the small changes and not the changes pt 2. The same issues are there but the small thing like getting ammo to spawn and the explosion not working anymore is fixed.

Jiymeade commented:
Sorry for all the comments, but I fixed the visible circles issue for the TNTSpawners

Jiymeade committed:
Update

-I believe I fixed the TNT spawner

MRN-DX commented:
![image](https://github.com/user-attachments/assets/c6af1dc3-a2b8-46b5-a914-3ccf4f5d7a3c)
you want to make sure there's only one script per object, if they spawn at the same time they, since they use the relative transform point of the game object, they will instantly collide.  I would also move the spawning a little bit higher over the ground plane

MRN-DX commented:
to make it less "watery" its because the mass of the ground plane is 1, make it heavier

![image](https://github.com/user-attachments/assets/26cffd36-8bb6-4a89-b990-f11b022e5f95)

MRN-DX commented:
Bullet Prefab is missing script and needs some tweaking 
![image](https://github.com/user-attachments/assets/cdf09fcf-1b73-4a7f-88c2-a3e91f0256e5)

MRN-DX commented:
You've tagged 3 Prefabs with the tag "Player", considering our logic with collision is dertermined around these tags we want to ensure only players have that tag, right now you are getting null exceptions because all the interactive objects are trying to pull non-existent bullet spawners as it trys to add ammo or do damage, etc. You need to remove these tags on these prefabs

![image](https://github.com/user-attachments/assets/730a86d4-1791-46d0-9940-be079c147f8d)

If we want the Projectiles to work with the TNT script, 
![image](https://github.com/user-attachments/assets/1163f714-47c2-4baa-b4c0-cc58f3a2b768)
We need to make a new tag for bullet and tag that with "Projectile"

MRN-DX commented:
The version I pulled still had a typo, make sure its lowercase like this, or double check if it matches with the animator it talks to
![image](https://github.com/user-attachments/assets/6993bb43-901e-4cf6-95d6-0792d1cfbdc0)

MRN-DX commented:
I believe after all those changes and tweaks you should have most of the core features done.

Jiymeade committed:
Final Update

- The my win text still doesn't show on the screen but everything else works

MRN-DX commented:
> Bullet Prefab is missing script and needs some tweaking ![image](https://private-user-images.githubusercontent.com/129788521/392971798-cdf09fcf-1b73-4a7f-88c2-a3e91f0256e5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM2ODgwMjQsIm5iZiI6MTczMzY4NzcyNCwicGF0aCI6Ii8xMjk3ODg1MjEvMzkyOTcxNzk4LWNkZjA5ZmNmLTFiNzMtNGE3Zi04OGMyLWEzZTkxZjAyNTZlNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwOFQxOTU1MjRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lNjJhNDgxNWEwNTM2OTVkMWNkZTgyMjA1OWUxODk4NmFkZDM1NDViYzNiYWYxYTZmOGRlN2FkZjBkZWM3YWFkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.2ZFhnRMzob77CJog6FycIdSY0JgxQPiMwJIWm93HZjY)

Still missing projectile script from bullet prefab, bullets cannot work without this script in the main prefab.

MRN-DX commented on UC NetCode Fall 2024/Assets/App/Resource/Scripts/GameScript.cs:
```diff
@@ -0,0 +1,70 @@
+
+        [Rpc(target:SendTo.Server)]
+
+        public void AddPlayerRpc(RpcParams rpcParams = default)
```
The reason why it's not "working" is because we are not calling this in the HealthScript

MRN-DX commented on UC NetCode Fall 2024/Assets/App/Resource/Scripts/NetHelpers/HealthNetScript.cs:
```diff
@@ -0,0 +1,71 @@
+
+            _Health.Value = _startingHealth;
+
+            _Health.OnValueChanged += UpdateHealth;
```
We need to tell the gameScript we are an active player, by calling `FindObjectOfType<GameScript>().AddPlayerRpc();`  here in the OnNetworkSpawn override.

MRN-DX commented on UC NetCode Fall 2024/Assets/App/Resource/Scripts/TntScript.cs:
```diff
@@ -0,0 +1,46 @@
+            {
+                TriggerExplosionRpc();
+            }
+        }
```
Right now the Projectile and explosion tag only matter to the TnT script, so its important to make sure those are matching accordingly.

now in your Bullet Prefab you have it set to a tag of Bullet. You would need to update this OncollisionEnter script if you were trying to get it to work with this current tag to `other.gameObject.tag.Equals("Bullet")`
![image](https://github.com/user-attachments/assets/cd93d8b8-bb3b-453f-875f-8cd95ce28270)

MRN-DX reviewed (changes requested):
50 / 60 now. You are so close its silly, if you want to make another commit you can.
-The biggest thing is the game script are not set correctly in the healthScript
-bullets not configured correctly.
